### PR TITLE
[OptimizeInstructions] Combine some relational ops joined Or/And (Part 7-8)

### PR DESCRIPTION
**Pull request by MaxGraey**
This is final part of #4265

```rust
(i32(x) >= 0) & (i32(y) >= 0)   ==>   i32(x | y) >= 0
(i64(x) >= 0) & (i64(y) >= 0)   ==>   i64(x | y) >= 0

(i32(x) == -1) & (i32(y) == -1)   ==>   i32(x & y) == -1
(i64(x) == -1) & (i64(y) == -1)   ==>   i64(x & y) == -1
```

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed: `ITERATION: 11267`

@kripken could you take a look please?

File: test/lit/passes/optimize-instructions.wast [hunk 1]
```diff
@@ -11253,6 +11253,108 @@
...
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $optimize-combined-by-and-greatequal-zero (param $x i32) (param $y i32) (param $a i64) (param $b i64)
+    ;; (i32(x) >= 0) & (i32(y) >= 0)   ==>   i32(x | y) >= 0
+    (drop (i32.and
+      (i32.ge_s (local.get $x) (i32.const 0))
+      (i32.ge_s (local.get $y) (i32.const 0))
+    ))
+    ;; (i64(x) >= 0) & (i64(y) >= 0)   ==>   i64(x | y) >= 0
+    (drop (i32.and
```
Comment (kripken) [inline test/lit/passes/optimize-instructions.wast:11295]: Shouldn't this be `i64`? Does binaryen not error on this?

Comment (MaxGraey) [reply test/lit/passes/optimize-instructions.wast:11295]: No, result type of LHS and RHS comparisons are `i32`. Also all this optimizations only apply for i32 and / or.

Comment (kripken) [reply test/lit/passes/optimize-instructions.wast:11295]: Oh, thanks, I forgot that the comparisons return a bool, which is `i32`...

**Review by kripken — APPROVED** [review]

**Pull request merged**